File: javascript/min_max_spec.js
var should = require('chai').should();
var Algos = require('./min_max');

var Ary = function () {
  return [2, 3, 1, 4, 5];
};

describe('min/max algorithms', function(){
  
  it('min', function(){
    Algos.min(2, 3).should.be.equal(2);
  });

  it('max', function(){
    Algos.max(2, 3).should.be.equal(3);
  });

  it('minmax', function(){
    Algos.minmax(2, 3).min.should.be.equal(2);
    Algos.minmax(2, 3).max.should.be.equal(3);
  });

  it('max_element', function(){
    Algos.max_element(Ary()).should.be.equal(5);
  });

  it('min_element', function(){
    Algos.min_element(Ary()).should.be.equal(1);
  });

  it('minmax_element', function(){
    Algos.minmax_element(Ary()).max_element.should.be.equal(5);
    Algos.minmax_element(Ary()).min_element.should.be.equal(1);
  });

  it('lexicographical_compare', function(){
    var a = 'abcd';
    var b = 'abce';
    var c = 'abcde';
    Algos.lexicographical_compare(a, b).should.be.equal(true);
    Algos.lexicographical_compare(c, a).should.be.equal(false);
    Algos.lexicographical_compare(a, c).should.be.equal(true);
  });

  it('is_permutation', function(){
    Algos.is_permutation(Ary(), [1, 2, 3, 4, 5]).should.be.equal(true);
    Algos.is_permutation(Ary(), [1, 2, 2, 4, 5]).should.be.equal(false);
    Algos.is_permutation(Ary(), [1, 2, 3, 4]).should.be.equal(false);
  });

  it('next_permutation', function(){
    var a = Ary();
    Algos.next_permutation(a);
    JSON.stringify(a).should.be.equal(JSON.stringify([2, 3, 1, 5, 4]));
  });

  it('prev_permutation', function(){
    var a = Ary();
    Algos.next_permutation(a);
    Algos.prev_permutation(a);
    JSON.stringify(a).should.be.equal(JSON.stringify(Ary()));
  });
});

File: lua/min_max_spec.lua
local Algos = require('./min_max')

local ary = {2, 3, 1, 4, 5}

describe('min/max algorithms', function()
    it('min', function()
        assert.are.same(Algos.min(2, 3), 2)
    end)

    it('max', function()
        assert.are.same(Algos.max(2, 3), 3)
    end)

    it('minmax', function()
        local min, max = Algos.minmax(2, 3)
        assert.are.same(min, 2)
        assert.are.same(max, 3)
    end)

    it('maxElement', function()
        assert.are.same(Algos.maxElement(ary), 5)
    end)

    it('minElement', function()
        assert.are.same(Algos.minElement(ary), 1)
    end)

    it('minmaxElement', function()
        local min, max = Algos.minmaxElement(ary)
        assert.are.same(min, 1)
        assert.are.same(max, 5)
    end)

    it('lexicographicalCompare', function()
        local a = 'abcd'
        local b = 'abce'
        local c = 'abced'
        assert.is_true(Algos.lexicographicalCompare(a, b))
        assert.is_true(Algos.lexicographicalCompare(b, c))
        assert.is_false(Algos.lexicographicalCompare(c, b))
    end)

    it('isPermutation', function()
        assert.is_true(Algos.isPermutation(ary, {1, 2, 3, 4, 5}))
        assert.is_false(Algos.isPermutation(ary, {1, 2, 3, 4, 4}))
        assert.is_false(Algos.isPermutation(ary, {1, 2, 3, 4}))
    end)

    it('nextPermutation', function()
        local list = {2, 3, 1, 4, 5}
        -- permutate table in-place
        Algos.nextPermutation(list)
        assert.are.same(list, {2, 3, 1, 5, 4})
    end)

    it('prevPermutation', function()
        local list = {2, 3, 1, 4, 5}
        -- assume nextPermutation is correct
        Algos.nextPermutation(list)
        Algos.prevPermutation(list)
        assert.are.same(list, {2, 3, 1, 4, 5})
    end)
end)
